{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-06-13 12:15:10.484066: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0\n"
     ]
    }
   ],
   "source": [
    "import importlib\n",
    "from pathlib import Path\n",
    "import itertools\n",
    "import pickle\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import tensorflow as tf\n",
    "from tensorflow.keras import backend as K\n",
    "from tensorflow.keras.losses import mean_absolute_error\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "import scipy.stats\n",
    "from statsmodels import api as sm\n",
    "import pyBigWig\n",
    "import logomaker\n",
    "import json\n",
    "\n",
    "from Modules import utils, plot_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "importlib.reload(utils)\n",
    "importlib.reload(plot_utils)\n",
    "data_dir = '../shared_folder'\n",
    "writing_dir = '../shared_folder'\n",
    "# writing_dir='../'\n",
    "prop_cycle = plt.rcParams['axes.prop_cycle']\n",
    "colors = prop_cycle.by_key()['color']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def correlate(y_true, y_pred):\n",
    "    \"\"\"\n",
    "\t\tCalculate the correlation between the predictions and the labels.\n",
    "\n",
    "\t\t:Example:\n",
    "\n",
    "\t\t>>> model.compile(optimizer = 'adam', losses = correlate)\n",
    "\t\t>>> load_model('file', custom_objects = {'correlate : correlate})\n",
    "\t\"\"\"\n",
    "    X = y_true - K.mean(y_true)\n",
    "    Y = y_pred - K.mean(y_pred)\n",
    "    \n",
    "    sigma_XY = K.sum(X*Y)\n",
    "    sigma_X = K.sqrt(K.sum(X*X))\n",
    "    sigma_Y = K.sqrt(K.sum(Y*Y))\n",
    "    \n",
    "    return sigma_XY/(sigma_X*sigma_Y + K.epsilon())\n",
    "\n",
    "def mae_cor(y_true, y_pred):\n",
    "    \"\"\"\n",
    "\t   Calculate the mean absolute error minus the correlation between\n",
    "        predictions and  labels.\n",
    "\n",
    "\t\t:Example:\n",
    "\n",
    "\t\t>>> model.compile(optimizer = 'adam', losses = mae_cor)\n",
    "\t\t>>> load_model('file', custom_objects = {'mae_cor : mae_cor})\n",
    "\t\"\"\"\n",
    "    X = y_true - K.mean(y_true)\n",
    "    Y = y_pred - K.mean(y_pred)\n",
    "    \n",
    "    sigma_XY = K.sum(X*Y)\n",
    "    sigma_X = K.sqrt(K.sum(X*X))\n",
    "    sigma_Y = K.sqrt(K.sum(Y*Y))\n",
    "    \n",
    "    cor = sigma_XY/(sigma_X*sigma_Y + K.epsilon())\n",
    "    mae = K.mean(K.abs(y_true - y_pred))\n",
    "    \n",
    "    return 1 + mae - cor\n",
    "\n",
    "def rolling_window(array, window=(0,), asteps=None, wsteps=None, axes=None, toend=True):  \n",
    "    \"\"\" \n",
    "        Take a numpy array and return a view of this array after applying a rolling window.\n",
    "\n",
    "        This takes a numpy and cut it in several pieces with the size, the stride and the \n",
    "        axes asked as needed. You may want to use it in order to create a set of sequences\n",
    "        from an array.\n",
    "\n",
    "        :param array: The array we want to cut\n",
    "        :param window: The length of the window\n",
    "        :param asteps: The stride between two window applied\n",
    "        :param wsteps: The stride whitin the window\n",
    "        :param axes: The axe on which to apply the rolling window\n",
    "        :param toend: Weither or not to finish the cut\n",
    "        :type array: numpy array\n",
    "        :type window: int or tuple\n",
    "        :type asteps: int or tuple\n",
    "        :type wsteps: int or tuple\n",
    "        :type axes: int\n",
    "        :type toend: boolean\n",
    "        :return: The view of the array\n",
    "        :rtype: numpy array\n",
    "\n",
    "        :Example:\n",
    "\n",
    "        >>> a = numpy.array([0,1,2,3,4,5])\n",
    "        >>> rolling_window(a, window = 2, asteps = 2, wsteps = None)\n",
    "        array([[0,1],\n",
    "               [2,3],\n",
    "               [4,5]])\n",
    "        >>> rolling_window(a, window = 2, asteps = None, wsteps = 2)\n",
    "        array([[0,2],\n",
    "               [1,3],\n",
    "               [2,4]\n",
    "               [3,5]])\n",
    "        >>> rolling_window(a, window = 5, asteps = 2, wsteps = None)\n",
    "        array([[0,1,2,3,4]])\n",
    "\n",
    "        .. warning:: Be carreful about the combination of window, wsteps and asteps that may raise \n",
    "                     ValueError. This function forces the window to be of the asked size and thus \n",
    "                     may stop the application of the window before the end.\n",
    "    \"\"\"        \n",
    "\n",
    "    array = np.asarray(array)\n",
    "    orig_shape = np.asarray(array.shape)\n",
    "    window = np.atleast_1d(window).astype(int) # maybe crude to cast to int...\n",
    "\n",
    "    if axes is not None:\n",
    "        axes = np.atleast_1d(axes)\n",
    "        w = np.zeros(array.ndim, dtype=int)\n",
    "        for axis, size in zip(axes, window):\n",
    "            w[axis] = size\n",
    "        window = w\n",
    "\n",
    "    # Check if window is legal:\n",
    "    if window.ndim > 1:\n",
    "        raise ValueError(\"`window` must be one-dimensional.\")\n",
    "    if np.any(window < 0):\n",
    "        raise ValueError(\"All elements of `window` must be larger than 1.\")\n",
    "    if len(array.shape) < len(window):\n",
    "        raise ValueError(\"`window` length must be less or equal `array` dimension.\") \n",
    "\n",
    "    _asteps = np.ones_like(orig_shape)\n",
    "    if asteps is not None:\n",
    "        asteps = np.atleast_1d(asteps)\n",
    "        if asteps.ndim != 1:\n",
    "            raise ValueError(\"`asteps` must be either a scalar or one dimensional.\")\n",
    "        if len(asteps) > array.ndim:\n",
    "            raise ValueError(\"`asteps` cannot be longer then the `array` dimension.\")\n",
    "        # does not enforce alignment, so that steps can be same as window too.\n",
    "        _asteps[-len(asteps):] = asteps\n",
    "        \n",
    "        if np.any(asteps < 1):\n",
    "             raise ValueError(\"All elements of `asteps` must be larger then 1.\")\n",
    "    asteps = _asteps\n",
    "\n",
    "    _wsteps = np.ones_like(window)\n",
    "    if wsteps is not None:\n",
    "        wsteps = np.atleast_1d(wsteps)\n",
    "        if wsteps.shape != window.shape:\n",
    "            raise ValueError(\"`wsteps` must have the same shape as `window`.\")\n",
    "        if np.any(wsteps < 0):\n",
    "             raise ValueError(\"All elements of `wsteps` must be larger then 0.\")\n",
    "\n",
    "        _wsteps[:] = wsteps\n",
    "        _wsteps[window == 0] = 1 # make sure that steps are 1 for non-existing dims.\n",
    "    wsteps = _wsteps\n",
    "\n",
    "    # Check that the window would not be larger than the original:\n",
    "    if np.any(orig_shape[-len(window):] < window * wsteps):\n",
    "        raise ValueError(\"`window` * `wsteps` larger then `array` in at least one dimension.\")\n",
    "\n",
    "    new_shape = orig_shape # just renaming...\n",
    "\n",
    "    # For calculating the new shape 0s must act like 1s:\n",
    "    _window = window.copy()\n",
    "    _window[_window==0] = 1\n",
    "\n",
    "    new_shape[-len(window):] += wsteps - _window * wsteps\n",
    "    new_shape = (new_shape + asteps - 1) // asteps\n",
    "    # make sure the new_shape is at least 1 in any \\\"old\\\" dimension (ie. steps\n",
    "    # is (too) large, but we do not care.\n",
    "    new_shape[new_shape < 1] = 1\n",
    "    shape = new_shape\n",
    "\n",
    "    strides = np.asarray(array.strides)\n",
    "    strides *= asteps\n",
    "    new_strides = array.strides[-len(window):] * wsteps\n",
    "    \n",
    "    # The full new shape and strides:\n",
    "    if toend:\n",
    "        new_shape = np.concatenate((shape, window))\n",
    "        new_strides = np.concatenate((strides, new_strides))\n",
    "    else:\n",
    "        _ = np.zeros_like(shape)\n",
    "        _[-len(window):] = window\n",
    "        _window = _.copy()\n",
    "        _[-len(window):] = new_strides\n",
    "        _new_strides = _\n",
    "        \n",
    "        new_shape = np.zeros(len(shape)*2, dtype=int)\n",
    "        new_strides = np.zeros(len(shape)*2, dtypenucleotid=int)\n",
    "        \n",
    "        new_shape[::2] = shape\n",
    "        new_strides[::2] = strides\n",
    "        new_shape[1::2] = _window\n",
    "        new_strides[1::2] = _new_strides\n",
    "    \n",
    "    new_strides = new_strides[new_shape != 0]\n",
    "    new_shape = new_shape[new_shape != 0]\n",
    "    \n",
    "    return np.lib.stride_tricks.as_strided(array, shape=new_shape, strides=new_strides)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = 'SCerevisiae'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<function tensorflow.python.keras.losses.mean_absolute_error(y_true, y_pred)>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "model = tf.keras.models.load_model(Path(data_dir, species, 'models_etienne', 'weights_myco_rep1_cohesine.hdf5'), custom_objects={'correlate': correlate})\n",
    "model.loss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1621478/3145760620.py:86: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  labels_myco[key] = log(value / labels_ctrlmyco[key])\n",
      "/tmp/ipykernel_1621478/3145760620.py:86: RuntimeWarning: invalid value encountered in true_divide\n",
      "  labels_myco[key] = log(value / labels_ctrlmyco[key])\n",
      "/home/alex/anaconda3/envs/tf2.5/lib/python3.8/site-packages/numpy/lib/function_base.py:4486: RuntimeWarning: invalid value encountered in subtract\n",
      "  diff_b_a = subtract(b, a)\n",
      "/tmp/ipykernel_1621478/3145760620.py:90: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  labels_pneu[key] = log(value / labels_ctrlpneu[key])\n",
      "/tmp/ipykernel_1621478/3145760620.py:90: RuntimeWarning: invalid value encountered in true_divide\n",
      "  labels_pneu[key] = log(value / labels_ctrlpneu[key])\n"
     ]
    }
   ],
   "source": [
    "def log(x):\n",
    "    x_ = np.copy(x)\n",
    "    x_[np.isnan(x_)] = 0\n",
    "    # Replacing inf:\n",
    "    # Replace inf by max value\n",
    "    # x_[np.isinf(x_)] = np.max(x_[np.isfinite(x_)])\n",
    "    # Etienne's way\n",
    "    # x_[np.isinf(x_)] = np.percentile(x_, 99.5)\n",
    "    # x_[np.isinf(x_)] = np.percentile(x_, 90)\n",
    "    # Workaround Etienne's way to avoid inf in cohesine myco\n",
    "    q99_5, q98 = np.percentile(x_, [99.5, 98])\n",
    "    if np.isnan(q99_5) :\n",
    "        x_[np.isinf(x_)] = q98\n",
    "    else:\n",
    "        x_[np.isinf(x_)] = q99_5\n",
    "    # Compute log\n",
    "    x_ = np.log2(x_ + 1)\n",
    "    return x_\n",
    "\n",
    "def load_yeast_bw(filename, bact='myco', romantoint=False, process=None):\n",
    "    # Load bigwig\n",
    "    labels = utils.load_bw(filename)\n",
    "    # Change chromosome names\n",
    "    if romantoint:\n",
    "        for key in list(labels.keys()):\n",
    "            if key[:3] == 'chr':\n",
    "                new_key = 'chr' + format(utils.roman_to_int(key[3:]), '02d')\n",
    "                labels[new_key] = labels.pop(key)\n",
    "    if bact == 'myco':\n",
    "        chr_id = list({'Mmm_Linear_740kb.fa', 'Mmmyco'}.intersection(labels.keys()))\n",
    "        assert len(chr_id) == 1\n",
    "        labels['myco'] = labels.pop(chr_id[0])\n",
    "    elif bact == 'pneu':\n",
    "        labels['pneu'] = labels.pop('Pneu300')\n",
    "    # Remove unwanted chromosomes\n",
    "    for key in list(labels.keys()):\n",
    "        if key[:3] != 'chr' and key != bact:\n",
    "            labels.pop(key)\n",
    "    # Transform data\n",
    "    if process is not None:\n",
    "        if process == 'conv':\n",
    "            process = lambda x: utils.moving_average(x, 21, keepsize=True)\n",
    "        elif process == 'log':\n",
    "            process = lambda x: np.log2(1 + x)\n",
    "        for key, value in labels.items():\n",
    "            labels[key] = process(value)\n",
    "    # Normalize in CPM\n",
    "    thres = sum(v.sum() for v in labels.values()) / 1e6\n",
    "    for key, value in labels.items():\n",
    "        labels[key] = value / thres\n",
    "    return labels\n",
    "\n",
    "def load_yeast_nuc(mycofile=None, pneufile=None, asdict=True, priority='myco', process='conv'):\n",
    "    if mycofile is None:\n",
    "        mycofile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/Bacterial_genome_project/data/for_Julien/myco/coverage_RPKM_myco265.bw'\n",
    "    if pneufile is None:\n",
    "        pneufile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/Bacterial_genome_project/data/for_Julien/pneumo/coverage_RPKM_pneumo205.bw'\n",
    "    # Load bigwigs\n",
    "    labels_myco = load_yeast_bw(mycofile, bact='myco', process=process)\n",
    "    labels_pneu = load_yeast_bw(pneufile, bact='pneu', process=process)\n",
    "    # Get labels from priority file\n",
    "    labels = {}\n",
    "    if priority == 'myco':\n",
    "        for key, value in labels_myco.items():\n",
    "            labels[key] = value\n",
    "        labels['pneu'] = labels_pneu['pneu']\n",
    "    else:\n",
    "        for key, value in labels_pneu.items():\n",
    "            labels[key] = value\n",
    "        labels['myco'] = labels_myco['myco']\n",
    "    # Return dictionary by chromosome, or an array per genome\n",
    "    if asdict:\n",
    "        return labels\n",
    "    else:\n",
    "        return (\n",
    "            np.concatenate([v for k, v in labels.items() if k[:3] == 'chr']),\n",
    "            labels['myco'],\n",
    "            labels['pneu']\n",
    "        )\n",
    "\n",
    "def load_yeast_chip(labels_ipmyco, labels_ctrlmyco, labels_ippneu, labels_ctrlpneu,\n",
    "                    asdict=True, priority='myco'):\n",
    "    # Get log ratios\n",
    "    labels_myco = {}\n",
    "    for key, value in labels_ipmyco.items():\n",
    "        labels_myco[key] = log(value / labels_ctrlmyco[key])\n",
    "    thres_myco = sum(v.sum() for v in labels_myco.values()) / 1e6\n",
    "    labels_pneu = {}\n",
    "    for key, value in labels_ippneu.items():\n",
    "        labels_pneu[key] = log(value / labels_ctrlpneu[key])\n",
    "    thres_pneu = sum(v.sum() for v in labels_pneu.values()) / 1e6\n",
    "    # Get labels from priority file, and normalize in CPM\n",
    "    labels = {}\n",
    "    if priority == 'myco':\n",
    "        for key, value in labels_myco.items():\n",
    "            labels[key] = value / thres_myco\n",
    "        labels['pneu'] = labels_pneu['pneu'] / thres_pneu\n",
    "    else:\n",
    "        for key, value in labels_pneu.items():\n",
    "            labels[key] = value / thres_pneu\n",
    "        labels['myco'] = labels_myco['myco'] / thres_myco\n",
    "    # Return dictionary by chromosome, or an array per genome\n",
    "    if asdict:\n",
    "        return labels\n",
    "    else:\n",
    "        return (\n",
    "            np.concatenate([v for k, v in labels.items() if k[:3] == 'chr']),\n",
    "            labels['myco'],\n",
    "            labels['pneu']\n",
    "        )\n",
    "\n",
    "def load_yeast_coh(mycoipfile=None, mycoctrlfile=None, pneuipfile=None, pneuctrlfile=None,\n",
    "                   asdict=True, priority='pneu'):\n",
    "    if mycoipfile is None:\n",
    "        mycoipfile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/GSM6703630_Scc1_Mmyco_rep1.CPM.bw'\n",
    "    if mycoctrlfile is None:\n",
    "        mycoctrlfile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/GSM6703631_Scc1-inp_Mmyco_rep1.CPM.bw'\n",
    "    if pneuipfile is None:\n",
    "        pneuipfile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/Bacterial_genome_project/data/for_Julien/pneumo/coverage_RPKM219.bw'\n",
    "    if pneuctrlfile is None:\n",
    "        pneuctrlfile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/Bacterial_genome_project/data/for_Julien/pneumo/coverage_RPKM220.bw'\n",
    "    # Load bigwigs\n",
    "    labels_ipmyco = load_yeast_bw(mycoipfile, bact='myco', romantoint=True)\n",
    "    labels_ctrlmyco = load_yeast_bw(mycoctrlfile, bact='myco', romantoint=True)\n",
    "    labels_ippneu = load_yeast_bw(pneuipfile, bact='pneu')\n",
    "    labels_ctrlpneu = load_yeast_bw(pneuctrlfile, bact='pneu')\n",
    "    return load_yeast_chip(labels_ipmyco, labels_ctrlmyco, labels_ippneu, labels_ctrlpneu,\n",
    "                           asdict=asdict, priority=priority)\n",
    "\n",
    "def load_yeast_pol(mycoipfile=None, mycoctrlfile=None, pneuipfile=None, pneuctrlfile=None,\n",
    "                   asdict=True, priority='myco'):\n",
    "    if mycoipfile is None:\n",
    "        mycoipfile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/Bacterial_genome_project/data/for_Julien/myco/coverage_RPKM102.bw'\n",
    "    if mycoctrlfile is None:\n",
    "        mycoctrlfile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/Bacterial_genome_project/data/for_Julien/myco/coverage_RPKM105.bw'\n",
    "    if pneuipfile is None:\n",
    "        pneuipfile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/Bacterial_genome_project/data/for_Julien/pneumo/coverage_RPKM154.bw'\n",
    "    if pneuctrlfile is None:\n",
    "        pneuctrlfile = '/home/alex/shared_folder/SCerevisiae/data/Data_etienne/Bacterial_genome_project/data/for_Julien/pneumo/coverage_RPKM155.bw'\n",
    "    # Load bigwigs\n",
    "    labels_ipmyco = load_yeast_bw(mycoipfile, bact='myco')\n",
    "    labels_ctrlmyco = load_yeast_bw(mycoctrlfile, bact='myco')\n",
    "    labels_ippneu = load_yeast_bw(pneuipfile, bact='pneu')\n",
    "    labels_ctrlpneu = load_yeast_bw(pneuctrlfile, bact='pneu')\n",
    "    return load_yeast_chip(labels_ipmyco, labels_ctrlmyco, labels_ippneu, labels_ctrlpneu,\n",
    "                           asdict=asdict, priority=priority)\n",
    "\n",
    "# labels_nuc_yeast, labels_nuc_myco, labels_nuc_pneu = load_yeast_nuc(asdict=False, process=None)\n",
    "# labels_coh_yeast, labels_coh_myco, labels_coh_pneu = load_yeast_coh(asdict=False)\n",
    "# labels_pol_yeast, labels_pol_myco, labels_pol_pneu = load_yeast_pol(asdict=False)\n",
    "labels_nuc = load_yeast_nuc()\n",
    "labels_coh = load_yeast_coh()\n",
    "labels_pol = load_yeast_pol()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54.476128 1.0072502 0.68775374\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'chr01': 0.49261805,\n",
       " 'chr02': 0.5665447,\n",
       " 'chr03': 0.5001005,\n",
       " 'chr04': 0.4937488,\n",
       " 'chr05': 0.6099528,\n",
       " 'chr06': 0.50966746,\n",
       " 'chr07': 0.5758628,\n",
       " 'chr08': 0.62615305,\n",
       " 'chr09': 0.47404206,\n",
       " 'chr10': 0.50311136,\n",
       " 'chr11': 0.67251915,\n",
       " 'chr12': 1.0072502,\n",
       " 'chr13': 0.51249945,\n",
       " 'chr14': 0.5261845,\n",
       " 'chr15': 0.5179289,\n",
       " 'chr16': 0.5935781,\n",
       " 'pneu': 0.49384093,\n",
       " 'myco': 0.5101825}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_nuc = np.max([np.max(v) for v in labels_nuc.values()])\n",
    "max_coh = np.max([np.max(v) for v in labels_coh.values()])\n",
    "max_pol = np.max([np.max(v) for v in labels_pol.values()])\n",
    "print(max_nuc, max_coh, max_pol)\n",
    "{k: np.max(v) for k, v in labels_nuc.items()}\n",
    "{k: np.max(v) for k, v in labels_coh.items()}\n",
    "# {k: np.max(v) for k, v in labels_pol.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='Count'>"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAD4CAYAAAD2FnFTAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8/fFQqAAAACXBIWXMAAAsTAAALEwEAmpwYAAAT40lEQVR4nO3df4xlZ33f8fcnY28BY3CSHYzx2qwRKxcnwmANtsEo2KGgNYGuqGi6FoUW4a5I7ShAQ+WkEpZa9S+iCIEdViuydWmxrRTsZJOsf1AF4TTE7o4d499G08XE03W7Yzu1wUR11vn2j3vWuZ19ZubuzJy5s7vvl3Q19zw/7nyfnZn72XPuueemqpAkab6fGncBkqT1yYCQJDUZEJKkJgNCktRkQEiSmk4adwGraePGjbV58+ZxlyFJx4x777336aqabPUdVwGxefNmpqenx12GJB0zkvxwoT4PMUmSmgwISVKTASFJajIgJElNBoQkqcmAkCQ19RYQSc5K8u0kjyZ5OMmvNcYkyZeSzCR5IMkFQ31bkzze9V3TV52SpLY+9yAOAf+qqt4CXAxcleS8eWMuB7Z0tx3AVwCSTADXd/3nAVc05kqSetTbG+Wq6ingqe7+j5I8CpwJPDI0bBvwtRp8KMXdSU5LcgawGZipqv0ASW7uxg7PXTUvvfQSMzMzL2+/+c1vZmJioo9vJUnHjDV5J3WSzcDbgXvmdZ0JPDm0Pdu1tdovWuCxdzDY++Dss89eVn0zMzPsuP6POWXjG3jh6QPsuuqXOPfcc5f1WJJ0vOj9Reokrwa+CXy6qp6f392YUou0H9lYtauqpqpqanKyeTmRkZyy8Q285vVv5JSNb1j2Y0jS8aTXPYgkJzMIh69X1S2NIbPAWUPbm4ADwIYF2iVJa6TPs5gC/C7waFX99gLD9gAf785muhh4rnvtYh+wJck5STYA27uxkqQ10ucexCXAx4AHk9zftf0mcDZAVe0E9gIfAGaAnwCf6PoOJbkauAOYAHZX1cM91ipJmqfPs5j+G+3XEobHFHDVAn17GQSIJGkMfCe1JKnJgJAkNRkQkqQmA0KS1GRASJKaDAhJUpMBIUlqMiAkSU0GhCSpyYCQJDUZEJKkJgNCktRkQEiSmgwISVKTASFJajIgJElNvX1gUJLdwAeBg1X1843+zwEfHarjLcBkVT2b5AngR8BLwKGqmuqrTklSW597EDcAWxfqrKovVNXbquptwG8A36mqZ4eGXNb1Gw6SNAa9BURV3QU8u+TAgSuAm/qqRZJ09Mb+GkSSVzHY0/jmUHMBdya5N8mO8VQmSSe23l6DOAofAv5s3uGlS6rqQJLXAd9K8li3R3KELkB2AJx99tn9VytJJ4ix70EA25l3eKmqDnRfDwK3AhcuNLmqdlXVVFVNTU5O9lqoJJ1IxhoQSV4LvAf4g6G2U5Kcevg+8H7gofFUKEknrj5Pc70JuBTYmGQWuBY4GaCqdnbDPgzcWVUvDE09Hbg1yeH6bqyq2/uqU5LU1ltAVNUVI4y5gcHpsMNt+4Hz+6lKkjSq9fAahCRpHTIgJElNBoQkqcmAkCQ1GRCSpCYDQpLUZEBIkpoMCElSkwEhSWoyICRJTQaEJKnJgJAkNRkQkqQmA0KS1GRASJKaDAhJUpMBIUlq6i0gkuxOcjBJ8/Okk1ya5Lkk93e3zw/1bU3yeJKZJNf0VaMkaWF97kHcAGxdYsyfVtXbutu/BUgyAVwPXA6cB1yR5Lwe65QkNfQWEFV1F/DsMqZeCMxU1f6qehG4Gdi2qsVJkpY07tcg3pnke0luS/JzXduZwJNDY2a7tqYkO5JMJ5mem5vrs1ZJOqGMMyDuA95YVecDXwZ+v2tPY2wt9CBVtauqpqpqanJycvWrlKQT1NgCoqqer6ofd/f3Aicn2chgj+GsoaGbgANjKFGSTmhjC4gkr0+S7v6FXS3PAPuALUnOSbIB2A7sGVedknSiOqmvB05yE3ApsDHJLHAtcDJAVe0EPgL8SpJDwF8D26uqgENJrgbuACaA3VX1cF91SpLaeguIqrpiif7rgOsW6NsL7O2jLknSaMZ9FpMkaZ0yICRJTQaEJKnJgJAkNRkQkqQmA0KS1GRASJKaDAhJUpMBIUlqMiAkSU0GhCSpyYCQJDUZEJKkJgNCktRkQEiSmgwISVKTASFJauotIJLsTnIwyUML9H80yQPd7btJzh/qeyLJg0nuTzLdV42SpIX1uQdxA7B1kf4fAO+pqrcC/w7YNa//sqp6W1VN9VSfJGkRfX4m9V1JNi/S/92hzbuBTX3VIkk6euvlNYhPArcNbRdwZ5J7k+xYbGKSHUmmk0zPzc31WqQknUh624MYVZLLGATEu4eaL6mqA0leB3wryWNVdVdrflXtojs8NTU1Vb0XLEkniLHuQSR5K/BVYFtVPXO4vaoOdF8PArcCF46nQkk6cY0tIJKcDdwCfKyqvj/UfkqSUw/fB94PNM+EkiT1p7dDTEluAi4FNiaZBa4FTgaoqp3A54GfBX4nCcCh7oyl04Fbu7aTgBur6va+6pQktfV5FtMVS/RfCVzZaN8PnH/kDEnSWlovZzFJktYZA0KS1GRASJKaRgqIJJeM0iZJOn6Mugfx5RHbJEnHiUXPYkryTuBdwGSSzw51vQaY6LMwSdJ4LXWa6wbg1d24U4fanwc+0ldRkqTxWzQgquo7wHeS3FBVP1yjmiRJ68Cob5T7e0l2AZuH51TVL/ZRlCRp/EYNiP8C7GRwYb2X+itHkrRejBoQh6rqK71WIklaV0Y9zfUPk/zLJGck+ZnDt14rkySN1ah7EP+s+/q5obYC3rS65UiS1ouRAqKqzum7EEnS+jJSQCT5eKu9qr62uuVIktaLUQ8xvWPo/iuA9wL3AQaEJB2nRj3E9KvD20leC/ynXiqSJK0Ly73c90+ALYsNSLI7ycEkzc+TzsCXkswkeSDJBUN9W5M83vVds8waJUkrMOprEH/I4KwlGFyk7y3A7y0x7QbgOhY+DHU5g5DZAlwEfAW4KMkEcD3wPmAW2JdkT1U9MkqtkqTVMeprEL81dP8Q8MOqml1sQlXdlWTzIkO2AV+rqgLuTnJakjMYXM5jpvtsapLc3I01ICRpDY10iKm7aN9jDK7o+tPAi6vwvc8Enhzanu3aFmpvSrIjyXSS6bm5uVUoS5IEo3+i3C8D/x34x8AvA/ckWenlvtNoq0Xam6pqV1VNVdXU5OTkCkuSJB026iGmfwO8o6oOAiSZBP4r8I0VfO9Z4Kyh7U3AAQafQdFqlyStoVHPYvqpw+HQeeYo5i5kD/Dx7mymi4HnquopYB+wJck5STYA27uxkqQ1NOoexO1J7gBu6rb/CbB3sQlJbgIuBTYmmQWuBU4GqKqd3fwPADMMTpv9RNd3KMnVwB0MzpjaXVUPH8WaJEmrYKnPpH4zcHpVfS7JPwLezeA1gj8Hvr7Y3Kq6Yon+Aq5aoG8vSwSQJKlfSx0m+iLwI4CquqWqPltVn2Hw5P3FfkuTJI3TUgGxuaoemN9YVdMM3q8gSTpOLRUQr1ik75WrWYgkaX1ZKiD2JfkX8xuTfBK4t5+SJEnrwVJnMX0auDXJR/m7QJhi8F6FD/dYlyRpzBYNiKr638C7klwG/HzX/MdV9Se9VyZJGqtRPw/i28C3e65FkrSOrPTd0JKk45QBIUlqMiAkSU0GhCSpyYCQJDUZEJKkJgNCktRkQEiSmgwISVKTASFJauo1IJJsTfJ4kpkk1zT6P5fk/u72UJKXkvxM1/dEkge7vuk+65QkHWnUz6Q+akkmgOuB9wGzDC4dvqeqHjk8pqq+AHyhG/8h4DNV9ezQw1xWVU/3VaMkaWF97kFcCMxU1f6qehG4Gdi2yPgrgJt6rEeSdBT6DIgzgSeHtme7tiMkeRWwFfjmUHMBdya5N8mOhb5Jkh1JppNMz83NrULZkiToNyDSaKsFxn4I+LN5h5cuqaoLgMuBq5L8QmtiVe2qqqmqmpqcnFxZxZKkl/UZELPAWUPbm4ADC4zdzrzDS1V1oPt6ELiVwSErSdIa6TMg9gFbkpyTZAODENgzf1CS1wLvAf5gqO2UJKcevg+8H3iox1olSfP0dhZTVR1KcjVwBzAB7K6qh5N8quvf2Q39MHBnVb0wNP10Bp+FfbjGG6vq9r5qlSQdqbeAAKiqvcDeeW07523fANwwr20/cH6ftUmSFuc7qSVJTQaEJKnJgJAkNRkQkqQmA0KS1GRASJKaDAhJUpMBIUlqMiAkSU0GhCSpyYCQJDUZEJKkJgNCktRkQEiSmgwISVKTASFJajIgJElNvQZEkq1JHk8yk+SaRv+lSZ5Lcn93+/yocyVJ/ertI0eTTADXA+8DZoF9SfZU1SPzhv5pVX1wmXMlST3pcw/iQmCmqvZX1YvAzcC2NZgrSVoFfQbEmcCTQ9uzXdt870zyvSS3Jfm5o5xLkh1JppNMz83NrUbdkiT6DYg02mre9n3AG6vqfODLwO8fxdxBY9WuqpqqqqnJycnl1ipJmqfPgJgFzhra3gQcGB5QVc9X1Y+7+3uBk5NsHGWuJKlffQbEPmBLknOSbAC2A3uGByR5fZJ09y/s6nlmlLmSpH71dhZTVR1KcjVwBzAB7K6qh5N8quvfCXwE+JUkh4C/BrZXVQHNuX3VKkk6Um8BAS8fNto7r23n0P3rgOtGnStJWju+k1qS1GRASJKaDAhJUpMBIUlqMiAkSU0GhCSpyYCQJDUZEJKkJgNCktRkQEiSmgwISVKTASFJajIgJElNBoQkqcmAkCQ1GRCSpCYDQpLU1GtAJNma5PEkM0muafR/NMkD3e27Sc4f6nsiyYNJ7k8y3WedkqQj9faRo0kmgOuB9wGzwL4ke6rqkaFhPwDeU1V/leRyYBdw0VD/ZVX1dF81SpIW1ucexIXATFXtr6oXgZuBbcMDquq7VfVX3ebdwKYe65EkHYU+A+JM4Mmh7dmubSGfBG4b2i7gziT3Jtmx0KQkO5JMJ5mem5tbUcGSpL/T2yEmII22ag5MLmMQEO8ear6kqg4keR3wrSSPVdVdRzxg1S4Gh6aYmppqPr4k6ej1uQcxC5w1tL0JODB/UJK3Al8FtlXVM4fbq+pA9/UgcCuDQ1aSpDXSZ0DsA7YkOSfJBmA7sGd4QJKzgVuAj1XV94faT0ly6uH7wPuBh3qsVZI0T2+HmKrqUJKrgTuACWB3VT2c5FNd/07g88DPAr+TBOBQVU0BpwO3dm0nATdW1e191SpJOlKfr0FQVXuBvfPadg7dvxK4sjFvP3D+/HZJ0trxndSSpCYDQpLUZEBIkpoMCElSkwEhSWoyICRJTQaEJKnJgJAkNRkQkqQmA0KS1GRASJKaDAhJUpMBIUlqMiAkSU0GhCSpyYCQJDUZEJKkpl4DIsnWJI8nmUlyTaM/Sb7U9T+Q5IJR50qS+tVbQCSZAK4HLgfOA65Ict68YZcDW7rbDuArRzFXktSjPj+T+kJgpvt8aZLcDGwDHhkasw34WlUVcHeS05KcAWweYe6qeuHpAy9/3b//tX19G0ladeeee24vj9tnQJwJPDm0PQtcNMKYM0ecC0CSHQz2PgB+nOTxZda7EXga4K4vL/MRjj0vr/kEcqKt+URbL7jmo/XGhTr6DIg02mrEMaPMHTRW7QJ2HV1pR0oyXVVTK32cY4lrPv6daOsF17ya+gyIWeCsoe1NwIERx2wYYa4kqUd9nsW0D9iS5JwkG4DtwJ55Y/YAH+/OZroYeK6qnhpxriSpR73tQVTVoSRXA3cAE8Duqno4yae6/p3AXuADwAzwE+ATi83tq9bOig9THYNc8/HvRFsvuOZVk8EJRJIk/f98J7UkqcmAkCQ1HfcBkWR3koNJHlqg/7i73Mdy15zkrCTfTvJokoeT/NraVr48K/kZd/0TSf4iyR+tTcUrt8Lf69OSfCPJY93P+p1rV/nyrXDNn+l+px9KclOSV6xd5cs3wpr/fpI/T/J/k/z6vL6VP39V1XF9A34BuAB4aIH+DwC3MXjvxcXAPV37BPA/gDcxOO32e8B5415Pz2s+A7igu38q8P1jYc3LXe9Q/2eBG4E/Gvda1mLNwH8EruzubwBOG/d6+lwzgzfe/gB4Zbf9e8A/H/d6VmnNrwPeAfx74NeH2lfl+eu434OoqruAZxcZ8vLlPqrqbuDw5T5evlRIVb0IHL7cx7q33DVX1VNVdV/3GD8CHmXwx7WureBnTJJNwC8BX+2/0tWz3DUneQ2DJ53f7R7nxar6P70XvApW8nNmcMbmK5OcBLyKY+R9VUutuaoOVtU+4G/mda3K89dxHxAjOJrLfaz7J8sRLbm2JJuBtwP3rF1ZvVlsvV8E/jXwt2tcU98WWvObgDngP3SH1b6a5JRxFNiD5pqr6n8CvwX8JfAUg/db3TmG+tbSqjx/GRCrcLmPY9Cia0vyauCbwKer6vk1q6o/zfUm+SBwsKruXeuC1sBCP+OTGByy+EpVvR14AThmXl9bwkI/559m8L/nc4A3AKck+adrWtnaW5XnLwNi4ct9jHKpkGPVgmtLcjKDcPh6Vd0yhtr6sNB6LwH+YZInGOyC/2KS/7z25fVisd/r2ao6vGf4DQaBcTxYaM3/APhBVc1V1d8AtwDvGkN9a2lVnr8MiBPzch/NNScJg2PTj1bVb4+3xFXVXG9V/UZVbaqqzQx+vn9SVcfL/ywXWvP/Ap5Mcvj60O+lx8vor7GF/pb/Erg4yau63/H3Mnh97Xi2Ks9ffV6sb11IchNwKbAxySxwLXAyrMvLfayK5a6Zwf+oPwY8mOT+ru03q2rvmhW/DCtY7zFrhWv+VeDr3RPHfo6Rf48V/C3fk+QbwH3AIeAvOEYux7HUmpO8HpgGXgP8bZJPMzhb6fnVeP7yUhuSpCYPMUmSmgwISVKTASFJajIgJElNBoQkqcmAkCQ1GRCSpKb/B7nnQTFpujL0AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.histplot(np.concatenate([np.clip(v, None, 1) for k, v in labels_coh.items() if k[:3] == 'chr']), bins=100, binrange=(1, 1.1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savez_compressed(Path(data_dir, species, 'models_etienne', 'labels_nuc.npz'), **labels_nuc)\n",
    "np.savez_compressed(Path(data_dir, species, 'models_etienne', 'labels_coh.npz'), **labels_coh)\n",
    "np.savez_compressed(Path(data_dir, species, 'models_etienne', 'labels_pol.npz'), **labels_pol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savez_compressed(Path(data_dir, species, 'models_etienne', 'labels_nuc_clipped.npz'), **{k: np.clip(v, None, 1) for k, v in labels_nuc.items()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with np.load(Path(data_dir, species, 'genome', 'W303', 'W303_ATCG.npz')) as f:\n",
    "    all_one_hots = {key: f[key] for key in f.keys() if key[:3] == 'chr'}\n",
    "with np.load(Path(data_dir, species, 'genome', 'bacterie', 'Mmm_Linear_740kb_ATCG.npz')) as f:\n",
    "    all_one_hots['myco'] = f['chr1']\n",
    "with np.load(Path(data_dir, species, 'genome', 'bacterie', 'Pneu300_ATCG.npz')) as f:\n",
    "    all_one_hots['pneu'] = f['chr1']\n",
    "np.savez(Path(data_dir, species, 'genome', 'bacterie', 'W303_myco_pneu_ATCG.npz'), **all_one_hots)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tf2.5",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
